tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -299,6 +302,33 @@ enum ArBasicKind {
 #define IS_BPROP_MIN_PRECISION(_Props) \
     (((_Props) & BPROP_MIN_PRECISION) != 0)
 
+#define IS_BPROP_16BIT_NO_MIN_PRECISION(_Props) \
+    (!IS_BPROP_MIN_PRECISION(_Props) && GET_BPROP_BITS(_Props) == BPROP_BITS16)
+
+#define IS_BPROP_16BIT_MIN_PRECISION(_Props) \
+    (IS_BPROP_MIN_PRECISION(_Props) && GET_BPROP_BITS(_Props) == BPROP_BITS16)
+
+#define IS_BPROP_16BIT_FLOAT_NO_MIN_PRECISION(_Props) \
+    (IS_BPROP_16BIT_NO_MIN_PRECISION(_Props) && IS_BPROP_FLOAT(_Props))
+
+#define IS_BPROP_16BIT_SINT_NO_MIN_PRECISION(_Props) \
+    (IS_BPROP_16BIT_NO_MIN_PRECISION(_Props) && IS_BPROP_SINT(_Props))
+
+#define IS_BPROP_16BIT_AINT_NO_MIN_PRECISION(_Props) \
```

Comment:
Comment, not sure what to do about it: This looks like it reads like slang: "ain't no min precision" (like: "you ain't no friend of mine" meaning "you are not a friend of mine").

With the recommended changes below, these new macros are not necessary and could be removed.